### ISSUE TITLE
dom-repeat template Indices/Keys not refreshed when an item is added-to/removed-from dom-repeat template data

### ISSUE THREAD

**Issue by shal1y** [Polymer/polymer#2270]
The dom-repeat template does not refresh the indices/keys when an item is added to the template data provider. It all works fine when adding the items in the end of the data (items) but when adding or removing items from any place "other" than the end, the template starts referring to the item just before or after the item being interacted with.
Check out the code snippet at:
http://plnkr.co/edit/U482WmTRhAAToump63cQ?p=preview

**Comment by jwu** [comment]
+1, also confirmed in my product.

**Comment by sorvell** [comment]
The link is down at the moment and I'm not able to confirm the problem based on the description. Hopefully the link will come back up, but any additional info would be helpful.

**Comment by jwu** [comment]
Here is a simple example to illustrate this problem:

```javascript
<html>

<head>
<title>Polymer Test</title>
<link rel="import" href="./bower_components/polymer/polymer.html">

<body>
    <template id="my-template" is="dom-bind">
        <button on-click="onAdd">add</button>
        <button on-click="onRemove">remove</button>
        <template is="dom-repeat" items="{{data}}">
            <div>
                <span>{{item.name}}</span>
                <button on-click="changeItem">change</button>
            </div>
        </template>
        <div id="info"></div>
    </template>

    <script>
    var tmpl = document.querySelector('#my-template');
    tmpl.data = [
        { name: 'Foobar 0' },
        { name: 'Foobar 1' },
        { name: 'Foobar 2' },
        { name: 'Foobar 3' },
    ]
    tmpl.onAdd = function () {
        tmpl.unshift('data', { name: 'Foobar ' + tmpl.data.length });
    }
    tmpl.onRemove = function () {
        tmpl.splice('data', tmpl.data.length-1, 1);
    }
    tmpl.changeItem = function ( event ) {
        event.model.set('item.name', event.model.item.name  + '#');
    }
    tmpl.addEventListener('data-changed', function ( event ) {
        tmpl.$.info.innerText = event.detail.path;
    });
    </script>

</body>

</html>
```

When you click `change` button for any of the item, you will see a log tell you who changes by path:

![screen shot 2015-08-15 at 10 36 25 am](https://cloud.githubusercontent.com/assets/174891/9287126/83bf8402-4339-11e5-821d-ed29121ef6e7.png)

After that you can click the `add` button, and the programme will add several item at the beginning of the array:

![screen shot 2015-08-15 at 10 37 26 am](https://cloud.githubusercontent.com/assets/174891/9287135/c3a0f5ba-4339-11e5-913f-c777671ec275.png)

So basically the array's indices should update by the new comers, if we click change button for `Foobar 6`, what I expect is a change path `data.0.name` but current Polymer will give me `data.6.name` which is wrong I guess.

![screen shot 2015-08-15 at 10 40 46 am](https://cloud.githubusercontent.com/assets/174891/9287141/1cc39c6a-433a-11e5-978d-d44b795328bc.png)

**Comment by azakus** [comment]
After investigating, this is a result of #2248, which is a high priority to fix

Here are some workarounds that you can use in the meantime
@shal1y 
http://plnkr.co/edit/gVsY7J2fsgo97ym2eubR?p=preview

@jwu 
http://jsbin.com/dogopakilu/1/edit?html,output

Closing in favor of #2248

**Comment by shal1y** [comment]
Thanks a lot @azakus

**Comment by jwu** [comment]
Thanks @azakus :)

**Comment by kevinpschaaf** [comment]
Closing as resolved, this was fixed somewhere along the way. No longer repros: https://lit.dev/playground/#gist=5b5f11c754f55829c7ead9f3b0a90a1e
